spark/src/main/scala/geotrellis/spark/knn/Implicits.scala

```diff
@@ -0,0 +1,10 @@
+package geotrellis.spark.knn
+
+import geotrellis.vector.Point
+import org.apache.spark.rdd.RDD
+
+object Implicits extends Implicits
+
+trait Implicits {
+  implicit class withKNearestMethods[T](val self: RDD[T]) extends KNearestMethods[T]
```

Comment:
Yeah, I agree, but I was following the example set by the SpatialIndex
implementation, which is generic over T and relies on the transformation T
=> Extent.  I thought it should be SpatialIndex[T: Geometry], but it
wasn't, so I didn't argue.  Providing the additional overloads should mimic
the proper functionality.  Of course, the better solution is just to fix
SpatialIndex.

On Wed, Jun 22, 2016 at 7:31 PM, Rob Emanuele <notifications@github.com>
wrote:

> In spark/src/main/scala/geotrellis/spark/knn/Implicits.scala
> <https://github.com/geotrellis/geotrellis/pull/1474#discussion_r68153986>:
>
> > @@ -0,0 +1,10 @@
> > +package geotrellis.spark.knn
> > +
> > +import geotrellis.vector.Point
> > +import org.apache.spark.rdd.RDD
> > +
> > +object Implicits extends Implicits
> > +
> > +trait Implicits {
> > +  implicit class withKNearestMethods[T](val self: RDD[T]) extends KNearestMethods[T]
>
> T is a bit generic. I can see, because we're saying, hey if you can give
> us a function that gives an extent then great, we can use this
> function...so perhaps that should stand. Unfortunate for compile times that
> this implicit class will be checked for every single call.
>
> I think we should make implicit methods for each of the known cases. E.g.
> I want to be able to call
>
> (rdd: RDD[PointFeature[D]]).kNearestNeighbors(point, 5)
> (rdd: RDD[Point]).kNearestNeighbors(point, 5)// etc...
>
> Also, we wanted an overload to call for a sequence of points (k nearest
> neighbors to the cluster centers):
>
> val clusterCenters: Traversable[Point] = ???val pointFeatures: RDD[PointFeature[T]] = ???val result: Map[Point, Seq[PointFeature[T]]] = pointFeatures.kNearestNeighbors(clusterCenters, 5)
> // for each key in result, there is a Seq of at most 5 points, which are the 5 closest neighbors to that cluster center
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotrellis/geotrellis/pull/1474/files/38d4ab856acd25cc23a9a71d9ebb31bddf2884ac#r68153986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ARjt0Xy3tkdhy92h1rX7Ka_1cL0Dnxrgks5qOcXRgaJpZM4IZJH0>
> .
>